yew/src/html/listener/registry.rs

```diff
@@ -0,0 +1,644 @@
+            let mut el = target;
+            loop {
+                el = match el.parent_element() {
+                    Some(el) => el,
+                    None => break,
+                };
+                run_handler(&el);
+            }
+        }
+    }
+}
+
+// Enable resetting between tests
+#[cfg(test)]
+impl Drop for Registry {
+    fn drop(&mut self) {
+        BODY.with(|body| {
+            for (kind, cl) in std::mem::take(&mut self.registered) {
+                AsRef::<web_sys::EventTarget>::as_ref(body)
+                    .remove_event_listener_with_callback(
+                        &kind.as_ref()[2..],
+                        cl.as_ref().unchecked_ref(),
+                    )
+                    .unwrap();
+            }
+        });
+    }
+}
+
+/// Register new event listeners to the element
+pub(crate) fn set_listeners(el: &Element, new: &mut Listeners) {
+    if let Listeners::Pending(ref mut pending) = new {
+        if pending.is_empty() {
+            return;
+        }
+
+        *new = Listeners::Registered(Registry::with(|reg| {
+            let id = reg.set_listener_id(el);
+            reg.register(id, std::mem::take(pending));
+            id
+        }));
+    }
+}
+
+/// Register new event listeners to the element, replacing old ones
+pub(crate) fn patch_listeners(mut new: &mut Listeners, old: &Listeners) {
+    if let (Listeners::Pending(pending), Listeners::Registered(id)) = (&mut new, old) {
+        Registry::with(|reg| {
+            // The listeners should never hav pointer equality so always replacing them is a slight
+            // optimisation
+            reg.unregister(id);
+            if pending.is_empty() {
+                return;
+            }
+            reg.register(*id, std::mem::take(pending));
+        });
+        *new = Listeners::Registered(*id);
+    }
+}
+
+/// Remove any registered event listeners from the element.
+/// The listener ID should be reused, if the element is reused.
+pub(crate) fn remove_listeners(listeners: &Listeners) {
+    if let Listeners::Registered(id) = listeners {
+        Registry::with(|reg| reg.unregister(id));
+    }
+}
+
+/// Compare passed listeners for equality with a registered set via pointer equality checks
+pub(crate) fn compare_listeners(registered_id: u32, rhs: &[Rc<dyn Listener>]) -> bool {
+    // Empty sets are not stored
+    if rhs.is_empty() {
+        return false;
+    }
+
+    Registry::with(|reg| match reg.by_id.get(&registered_id) {
+        Some(reg) => {
+            if reg.len() != rhs.len() {
+                return false;
+            }
+
+            rhs.iter()
+                .any(|l| match reg.get(&EventDescriptor::from(&**l)) {
+                    Some(reg) =>
+                    {
+                        #[allow(clippy::vtable_address_comparisons)]
+                        !Rc::ptr_eq(reg, l)
+                    }
+                    None => true,
+                })
```

Comment:
Are the booleans flipped here? Shouldn't this be checking that _all_ `rhs` listeners have a match? I think we'll have to switched to an ordered comparison anyways though since the order in which events are added is important

Reply:
Yes, I'll fix it.